### PR TITLE
MakeFile

### PR DESCRIPTION
**Pull request by layeghimahsa1996**
Hello everyone, 
I just pushed an initial version of Makefile. But it has a problem regarding sdl linkage. 
Can anyone help? I got stuck  ;<

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -1,2 +1,151 @@
-all:
-	gcc src/main.c src/user_manual.c src/is_reach_goal.c src/win_game.c src/print_menu.c src/credit.c src/single_player.c src/game_over.c src/take_heart.c src/is_player_hit.c src/is_barrier_hit.c src/load_map.c src/update_player_pos.c src/generate_view.c src/update_barrier.c src/update_arrow.c -l sdl2 -l sdl2_ttf -l sdl2_image
+CC=gcc 
+
+LIBFLAGS= -lmingw32 -lSDL2main -lSDL2
```
Comment (Pari-Firoozi) [inline Makefile:3]: Put an if statement like what we did for include SDL for different platform in src and include files.
for mac:
LIBFLAGS = -l sdl2 -l sdl2_ttf
for windows probably:
LIBFLAGS= -lmingw32 -lSDL2main -lSDL2

File: Makefile [hunk 2]
```diff
@@ -1,2 +1,151 @@
...
+
+#CREATE BIN AND BUILD FOLDERS TO SAVE THE COMPILED FILES DURING RUNTIME IF THEY DO NOT EXIST
+bin_folder := $(shell mkdir -p bin) 
+build_folder := $(shell mkdir -p build) 
+results_folder := $(shell mkdir -p program_output) 
+
+#TARGET TO GENERATE THE OBJECT FILES 
+main.o: src/main.c
```
Comment (Pari-Firoozi) [inline Makefile:11]: We need to put include files as prerequisites. You can define a variable as follows: 
INCLUDES = include/map.h include/menu.h include/single_player.h include/map_textures.h
and apply changes accordingly. For example:
build\main.o: src/main.c $(INCLUDES)

File: Makefile [hunk 3]
```diff
@@ -1,2 +1,151 @@
...
+	$(CC) -g -c $(LIBFLAGS) src/update_player_pos.c -o build/update_player_pos.o
+	
+user_manual.o: src/user_manual.c
+	$(CC) -g -c $(LIBFLAGS) src/user_manual.c -o build/user_manual.o
+	
+win_game.o: src/win_game.c
+	$(CC) -g -c $(LIBFLAGS) src/win_game.c -o build/win_game.o
+	
+	
+choose_menu_item_test.o: test/choose_menu_item_test.c
```
Comment (Pari-Firoozi) [inline Makefile:60]: Do the same for test files as well. Add includes and  change the ".o" files.
example:
bulid\choose_menu_item_test.o: test/choose_menu_item_test.c $(INCLUDES)

File: Makefile [hunk 4]
```diff
@@ -1,2 +1,151 @@
...
+	$(CC) -g -o bin/take_heart_test build/take_heart_test.o
+	$(CC) -g -o bin/update_arrow_test build/update_arrow_test.o
+	$(CC) -g -o bin/update_barrier_test build/update_barrier_test.o
+	$(CC) -g -o bin/update_player_pos_test build/update_player_pos_test.o
+	$(CC) -g -o bin/update_selector_test build/update_selector_test.o
+	$(CC) -g -o bin/win_game_test build/win_game_test.o
+
+
+	
+main: main.o credit.o game_over.o generate_view.o is_barrier_hit.o is_player_hit.o is_reach_goal.o load_map.o print_menu.o single_player.o take_heart.o update_arrow.o update_barrier.o update_player_pos.o user_manual.o win_game.o
```
Comment (Pari-Firoozi) [inline Makefile:126]: Define a variable as follows. The main rule just needs on recipe like what i did.
use the following:
OBJECTS = build/main.o build/credit.o build/game_over.o build/generate_view.o build/is_barrier_hit.o \
 		  build/is_player_hit.o build/is_reach_goal.o build/load_map.o build/print_menu.o \
 		  build/single_player.o build/take_heart.o build/update_arrow.o build/update_barrier.o \
 		  build/update_player_pos.o build/user_manual.o build/win_game.o
 		 
main : $(OBJECTS)
	$(CC) -o main $(OBJECTS) $(LIBFLAGS)

File: Makefile [hunk 5]
```diff
@@ -1,2 +1,151 @@
...
+
+	$(CC) -g -o bin/Run_and_Shoot build/main.o
```
Comment (Pari-Firoozi) [inline Makefile:128]: Remove line 128 to 143.

File: Makefile [hunk 6]
```diff
@@ -1,2 +1,151 @@
...
+	
+win_game_test.o: test/win_game_test.c
+	$(CC) -g -c $(LIBFLAGS) test/win_game_test.c -o build/win_game_test.o
+	
+
+
+
+#TARGET TO GENERATE THE EXECUTABLE OF THE PROGRAM (and the tests) - LINK OBJECT FILES
+
+tests: choose_menu_item_test.o game_over_test.o generate_view_test.o is_barrier_hit_test.o is_player_hit_test.o is_reach_goal_test.o load_map_test.o print_menu_test.o take_heart_test.o update_arrow_test.o update_barrier_test.o update_player_pos_test.o update_selector_test.o win_game_test.o
```
Comment (Pari-Firoozi) [inline Makefile:107]: Same comment as i put for main.

File: Makefile [hunk 7]
```diff
@@ -1,2 +1,151 @@
...
+	$(CC) -g -o bin/single_player single_player/main.o
+	$(CC) -g -o bin/take_heart build/take_heart.o
+	$(CC) -g -o bin/update_arrow build/update_arrow.o
+	$(CC) -g -o bin/update_barrier build/update_barrier.o
+	$(CC) -g -o bin/update_player_pos build/update_player_pos.o
+	$(CC) -g -o bin/user_manual build/user_manual.o
+	$(CC) -g -o bin/win_game build/win_game.o
+
+#TARGET TO GENERATE ALL THE EXECUTABLES (MAIN PROGRAM + TESTS TOGETHER)
+all: main tests
```
Comment (Pari-Firoozi) [inline Makefile:146]: I believe it should be the first rule to be executed, so put it on top of the file.

File: Makefile [hunk 8]
```diff
@@ -1,2 +1,151 @@
-all:
-	gcc src/main.c src/user_manual.c src/is_reach_goal.c src/win_game.c src/print_menu.c src/credit.c src/single_player.c src/game_over.c src/take_heart.c src/is_player_hit.c src/is_barrier_hit.c src/load_map.c src/update_player_pos.c src/generate_view.c src/update_barrier.c src/update_arrow.c -l sdl2 -l sdl2_ttf -l sdl2_image
+CC=gcc 
+
+LIBFLAGS= -lmingw32 -lSDL2main -lSDL2
+
+#CREATE BIN AND BUILD FOLDERS TO SAVE THE COMPILED FILES DURING RUNTIME IF THEY DO NOT EXIST
```
Comment (Pari-Firoozi) [inline Makefile:5]: I think we already created build and bin directories. You can either remove those two directory or these commands.
I think its better you remove bin and build directory and they are created at time of compilation like what you wrote here.

**Review by Pari-Firoozi — CHANGES_REQUESTED** [review]
Thank for youe effort Mahsa.
Please apply the comments.

File: Makefile [hunk 9]
```diff
@@ -1,2 +1,151 @@
...
+	$(CC) -g -c $(LIBFLAGS) src/game_over.c -o build/game_over.o
+	
+generate_view.o: src/generate_view.c
+	$(CC) -g -c $(LIBFLAGS) src/generate_view.c -o build/generate_view.o
+	
+is_barrier_hit.o: src/is_barrier_hit.c
+	$(CC) -g -c $(LIBFLAGS) src/is_barrier_hit.c -o build/is_barrier_hit.o
+
+is_player_hit.o: src/is_player_hit.c
+	$(CC) -g -c $(LIBFLAGS) src/is_player_hit.c -o build/is_barrier_hit.o
```
Comment (Pari-Firoozi) [inline Makefile:27]: $(CC) -g -c $(LIBFLAGS) src/is_player_hit.c -o build/is_player_hit.o

File: Makefile [hunk 10]
```diff
@@ -1,2 +1,151 @@
...
+	$(CC) -g -c $(LIBFLAGS) test/choose_menu_item_test.c -o build/choose_menu_item_test.o
+	
+game_over_test.o: test/game_over_test.c
+	$(CC) -g -c $(LIBFLAGS) test/game_over_test.c -o build/game_over_test.o
+
+generate_view_test.o: test/generate_view_test.c
+	$(CC) -g -c $(LIBFLAGS) test/generate_view_test.c -o build/generate_view_test.o
+
+is_barrier_hit_test.o: test/is_barrier_hit_test.c
+	$(CC) -g -c $(LIBFLAGS) test/is_barrier_hit_test.c -o build/is_barrier_hit.o
```
Comment (Pari-Firoozi) [inline Makefile:70]: $(CC) -g -c $(LIBFLAGS) test/is_barrier_hit_test.c -o build/is_barrier_hit_test.o

File: Makefile [hunk 11]
```diff
@@ -1,2 +1,151 @@
...
+	$(CC) -g -c $(LIBFLAGS) test/load_map_test.c -o build/load_map_test.o
+
+print_menu_test.o: test/print_menu_test.c
+	$(CC) -g -c $(LIBFLAGS) test/print_menu_test.c -o build/print_menu_test.o
+
+take_heart_test.o: test/take_heart_test.c
+	$(CC) -g -c $(LIBFLAGS) test/take_heart_test.c -o build/take_heart_test.o
+
+update_arrow_test.o: test/update_arrow_test.c
+	$(CC) -g -c $(LIBFLAGS) test/update_arrow_test.c -o build/update_arrow.o
```
Comment (Pari-Firoozi) [inline Makefile:88]: $(CC) -g -c $(LIBFLAGS) test/update_arrow_test.c -o build/update_arrow_test.o

**Review by Pari-Firoozi — CHANGES_REQUESTED** [review]

File: Makefile (hunk 4 above)
Comment (layeghimahsa1996) [reply Makefile:126]: That's much better!!!! Thanks for your comments on this.

File: Makefile (hunk 7 above)
Comment (layeghimahsa1996) [reply Makefile:146]: Yes You are completely right. I was curious to see main results first. I forgot to change the order :D

**Comment by layeghimahsa1996** [comment]
@Pari-Firoozi  @JeanneXu1122 @Jaserelhabrouk  
Can you review newly added files? I pushed makefile from ml_shoot_function branch by mistake.

File: src/update_bullet.c [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+	if (map->bullet[BULLET_1].speed != 0) { /**< We have a moving bullet shooted from player1*/
+		switch (map->bullet[BULLET_1].direction)
+		{
+		case DIRECTION_RIGHT:
+			if (map->bullet[BULLET_1].current_pos.x <= map->space.x_max)
+			{
+				map->bullet[BULLET_1].current_pos.x += map->bullet[BULLET_1].speed;
+			}
+			else {
+				map->bullet[BULLET_1].current_pos.x = 10000000;
```
Comment (Pari-Firoozi) [inline src/update_bullet.c:28]: So we will have lots of bullet hiding in the map somewhere :) and they should be processed each time.
If you define is_bullet_active as one of the player attribute you can switch it to be false and does not show that in generate view.

**Review by Pari-Firoozi — CHANGES_REQUESTED** [review]
Except  Makefile, the rest seems okay. Just it would be better if you apply the comments about the bullet.
Thanks.

File: include/map.h [hunk 13]
```diff
@@ -120,8 +137,9 @@ typedef struct map {
 	map_barrier_t barrier[MAP_MAX_NUM_OF_BARRIERS];    /**< barries in the map*/
 	arrow_t arrow[MAP_MAX_NUM_OF_ARROWS];              /**< arrows in the map*/
 	goal_t goal;                                       /**< goal in the map*/
-	player_t player;                                   /**< player in the map*/
+	player_t player[MAP_MAX_NUM_OF_PLAYERS];           /**< player in the map*/
 	map_textures_t textures;                           /**< map element textures*/
+	bullet_t bullet[MAP_MAX_NUM_OF_BULLETS];		   /**< bullet in the map*/
```
[the remark this answers is not in the document]
Comment (layeghimahsa1996) [reply include/map.h:142]: Very good suggestion. I will consider it. Thank you so much

File: src/update_bullet.c (hunk 12 above)
Comment (layeghimahsa1996) [reply src/update_bullet.c:28]: Yes you are right that's better solution indeed.

**Review by Pari-Firoozi — APPROVED** [review]

**Review by JeanneXu1122 — APPROVED** [review]
good

**Review by Pari-Firoozi — CHANGES_REQUESTED** [review]
You may want to close this pull request since it has been updated.

**Review by Jaserelhabrouk — APPROVED** [review]

**Pull request closed without merging**